File: src/main/java/algo/breadthFirstSearch/IntegerReplacement_397.java
package algo.breadthFirstSearch;

import java.util.ArrayDeque;

public class IntegerReplacement_397 {
  public static int integerReplacement(int n) {
    int ops = 0;
    var q = new ArrayDeque<Long>();
    q.offer((long) n);
    while (!q.isEmpty()) {
      var size = q.size();
      for (var i = 0; i < size; i++) {
        var e = q.poll();
        if (e == 1) {
          return ops;
        }
        if (e % 2 == 0) {
          q.offer(e / 2);
        } else {
          q.offer(e + 1);
          if (e > 1) {
            q.offer(e - 1);
          }
        }
      }
      ops++;
    }

    return -1;
  }

  public static void main(String[] args) {
//    System.out.println(integerReplacement(8));
//    System.out.println(integerReplacement(7));
    System.out.println(integerReplacement(2147483647));
  }
}

File: src/main/kotlin/algo/breadthFirstSearch/IntegerReplacement_397.kt
package algo.breadthFirstSearch

import java.util.LinkedList

class KtIntegerReplacement_397 {
    fun integerReplacement(n: Int): Int {
        var ops = 0
        val q = LinkedList<Long>()
        q.offer(n.toLong())
        while (q.isNotEmpty()) {
            val size = q.size
            for (i in 1..size) {
                val num = q.poll()
                if (num == 1L) {
                    return ops
                }
                when (num % 2) {
                    0L -> q.offer(num / 2)
                    1L -> {
                        q.offer(num - 1)
                        q.offer(num + 1)
                    }
                }
            }
            ops++
        }
        return -1
    }
}

fun main() {
    val replace = KtIntegerReplacement_397()
    println(replace.integerReplacement(8))
    println(replace.integerReplacement(7))
    println(replace.integerReplacement(4))
}